client/executor/src/integration_tests/mod.rs

```diff
@@ -775,3 +784,82 @@ fn panic_in_spawned_instance_panics_on_joining_its_result(wasm_method: WasmExecu
 
 	assert!(format!("{}", error_result).contains("Spawned task"));
 }
+
+
+#[cfg(target_os = "linux")]
+mod linux {
+	use super::*;
+
+	/// Gets the rollup of for the current process from procfs.
+	///
+	/// See the description here: https://www.kernel.org/doc/Documentation/ABI/testing/procfs-smaps_rollup
+	fn obtain_rss() -> u32 {
+		let smaps_rollup = std::fs::read("/proc/self/smaps_rollup").expect("failed to read smaps");
+		let smaps_rollup = String::from_utf8(smaps_rollup).expect("expected proper utf8");
+
+		for line in smaps_rollup.lines() {
+			// We are looking for a line that looks something like below. We are interested only in the
+			// number there.
+			//
+			// Rss:               63624 kB
+			//                    ^^^^^
+			//                    |
+			//                    = our target
+			//
+			// we don't even care about the suffix since it appears to be hardcoded to kB.
+
+			let line = match line.strip_prefix("Rss:") {
+				None => continue,
+				Some(line) => line,
+			};
+
+			let line = match line.strip_suffix("kB") {
+				None => {
+					panic!("weird, we expected that the smaps output is always terminated with `kB`");
+				}
+				Some(line) => line,
+			};
+
+			let line = line.trim();
+
+			let rss = line.parse::<u32>()
+				.expect("failed to parse the RSS");
+
+			return rss
+		}
+
+		panic!("smaps doesn't contain rss!");
+	}
+
+	test_wasm_execution!(compiled_only memory_consumption);
+	fn memory_consumption(wasm_method: WasmExecutionMethod) {
+		let runtime = mk_test_runtime(wasm_method, 1024);
+
+		let instance = runtime.new_instance().unwrap();
+		let heap_base = instance.get_global_const("__heap_base")
+			.expect("`__heap_base` is valid")
+			.expect("`__heap_base` exists")
+			.as_i32()
+			.expect("`__heap_base` is an `i32`");
+
+		instance.call_export("test_dirty_plenty_memory", &(heap_base as u32, 1u32).encode()).unwrap();
+
+		let rss_pre = obtain_rss();
+		instance.call_export("test_dirty_plenty_memory", &(heap_base as u32, 1024u32).encode()).unwrap();
+		let rss_post = obtain_rss();
+
+		// In case the memory is sucessfully decommited we expect only a slight increase of resident memory
+		// usage.
+		//
+		// However, in practice the memory usage can either go up and sometimes down (hence
+		// `saturating_sub` below). This is because this test is run along with other concurrently
+		// and the other tests may allocate or deallocate memory. This in theory makes this test
+		// flaky. This problem should be excaberated with the number of CPU cores
+		//
+		// However, empirically, this works just fine on the "build host" (numcpus=128) and is able
+		// reliably to tell if the decommiting really works or not.
+		//
+		// That said, if the test fails randomly we should consider turning it off.
+		assert!(rss_post.saturating_sub(rss_pre) < 32768 /* kB */);
```

Comment:
Wouldn't it be better to just look at the RSS of the specific pages where the WASM heap is kept instead of the whole process?

E.g. something like this test program:

```rust
use std::ops::Range;
use std::collections::BTreeMap;

fn get_smaps() -> Vec<(Range<usize>, BTreeMap<String, usize>)> {
    let regex_start = regex::RegexBuilder::new("^([0-9a-f]+)-([0-9a-f]+)").multi_line(true).build().unwrap();
    let regex_kv = regex::RegexBuilder::new(r#"^([^:]+):\s*(\d+) kB"#).multi_line(true).build().unwrap();
    let smaps = std::fs::read_to_string("/proc/self/smaps").unwrap();
    let boundaries: Vec<_> = regex_start
        .find_iter(&smaps)
        .map(|matched| matched.start())
        .chain(std::iter::once(smaps.len()))
        .collect();

    let mut output = Vec::new();
    for window in boundaries.windows(2) {
        let chunk = &smaps[window[0]..window[1]];
        let caps = regex_start.captures(chunk).unwrap();
        let start = usize::from_str_radix(caps.get(1).unwrap().as_str(), 16).unwrap();
        let end = usize::from_str_radix(caps.get(2).unwrap().as_str(), 16).unwrap();

        let values = regex_kv.captures_iter(chunk).map(|cap| {
            let key = cap.get(1).unwrap().as_str().to_owned();
            let value = cap.get(2).unwrap().as_str().parse().unwrap();
            (key, value)
        }).collect();

        output.push((start..end, values));
    }
                                                                                                                                      
    output                                                                                                                            
}                                                                                                                                     
                                                                                                                                      
fn get_map(addr: usize) -> BTreeMap<String, usize> {                                                                                  
    get_smaps().into_iter().find(|(range, _)| {                                                                                       
        addr >= range.start && addr < range.end                                                                                       
    }).unwrap().1                                                                                                                       
}

fn get_rss(addr: usize) -> usize {
    *get_map(addr).get("Rss").unwrap()
}

fn main() {
    unsafe {
        let p = libc::mmap(
            std::ptr::null_mut(),
            4096 * 16,
            libc::PROT_READ|libc::PROT_WRITE,
            libc::MAP_ANONYMOUS|libc::MAP_PRIVATE,
            0,
            0
        );

        let pp = p as *mut u8;
        println!("Initial RSS:");
        println!("{}", get_rss(pp as usize));

        println!("After dirtying one page:");
        pp.write(1);
        println!("{}", get_rss(pp as usize));

        println!("After dirtying two pages:");
        pp.add(4096).write(1);
        println!("{}", get_rss(pp as usize));

        println!("After MADV_DONTNEED:");
        libc::madvise(p, 4096 * 16, libc::MADV_DONTNEED);
        println!("{}", get_rss(pp as usize));
    }
}
```

Output:

```
Initial RSS:
0
After dirtying one page:
4
After dirtying two pages:
8
After MADV_DONTNEED:
0
```

Reply:
Yeah, good idea, that would be definitely more precise! It would require somehow plumbing the base addr of the linear memory here though. That sounds burdensome and not sure if it worth the effort right now IMO. OTOH, if this test proves to be flaky then that would be the way to go.

WDYT?

Reply:
Well, it's up to you. (:

Personally I think it's worthwhile *if* we want to have a test for this, since with the way the test is currently written its long-term value is somewhat questionable - you could both have it fail when it should succeed, and succeed when it should fail, so effectively does it *actually* protect us from a regression when it's not guaranteed to give a correct result?

But again, it's up to you. You know this code a lot better than I do after all!

Reply:
You convinced me.

See https://github.com/paritytech/substrate/pull/8998/commits/5d07a203269dab07a9404e7e5aecf08cdab40587 and later PRs.

The trait fn is annoying but it is not that much hassle.

Reply:
Looks good to me!